### PR TITLE
#96 - Test full order batches

### PR DESCRIPTION
**Pull request by bh2smith**
This test places 10 groups of 100 orders each. Thus filling two complete order batches (auction Ids 0 and 1). Afterwards one final order is placed and assertions are made that the auction index is 2.

### CODE REVIEW

File: test/snapp_auction.js [hunk 1]
```diff
@@ -520,4 +521,31 @@ contract("SnappAuction", async (accounts) => {
...
+
+  describe("Larger Test Cases", () => {
+    it("Fill order batch", async () => {
+      const instance = await SnappAuction.new()
+      await setupEnvironment(MintableERC20, instance, token_owner, [user_1], 2)
+      const order = encodeOrder(0, 1, 2, 3)
+      const maxAuctionSize = (await instance.AUCTION_BATCH_SIZE.call()).toNumber()
+      const orders = Array(maxAuctionSize).fill(order)
+      const partitionedOrders = splitArray(orders, 100)
+      let concatenated_orders
```
Comment (fleupold) [inline test/snapp_auction.js]: nit: I'd declare inside the forEach function.

File: test/snapp_auction.js [hunk 2]
```diff
@@ -520,4 +521,31 @@ contract("SnappAuction", async (accounts) => {
...
+      // let count = 0
+      await partitionedOrders.forEach(function(partition) {  // should be about 10
+        concatenated_orders = Buffer.concat(partition)
+        instance.placeSellOrders(concatenated_orders, { from: user_1 })
```
Comment (fleupold) [inline test/snapp_auction.js]: Don't you need to return an awaitable here for the top level await to take an effect?

File: test/snapp_auction.js [hunk 3]
```diff
@@ -520,4 +521,31 @@ contract("SnappAuction", async (accounts) => {
...
+      // const concatenated_orders = Buffer.concat(orders)
+      // await instance.placeSellOrders(concatenated_orders, { from: user_1 })
+ 
+      const auctionIndex = (await instance.auctionIndex.call()).toNumber()
+      const currentAuction = await instance.auctions(auctionIndex)
+      assert.equal(currentAuction.size.toNumber(), maxAuctionSize)
+ 
+      // The last order should wind up in second batch!
+      await instance.placeSellOrders(order, { from: user_1 })
+      // console.log((await instance.auctionIndex.call()).toNumber())
```
Comment (fleupold) [inline test/snapp_auction.js]: missing some assert? Also let's clean up all the commented code that's no longer used.

File: test/utilities.js [hunk 4]
```diff
@@ -147,6 +147,20 @@ const generateMerkleTree = memoize(_generateMerkleTree, {
   strategy: memoize.strategies.variadic
 })
 
+/**
+ * Partitions arrary into chunks of size spacing
+ * @param input: Array
+ * @param spacing: int
+ * @returns {Array}
+ */
+function splitArray(input, spacing) {
```
Comment (fleupold) [inline test/utilities.js]: nit: maybe call splitArrayIntoChunks to make it more clear what this function does without reading its docs?

ubernit: for those who love concise recursive (but not so easy to read) code:

```
return [input.slice(0, spacing)].concat(splitArray(input.slice(spacing), spacing));
```

**Review by fleupold — APPROVED** [review]
Nice.

🔪

File: test/snapp_auction.js (hunk 1 above)
Comment (bh2smith) [reply test/snapp_auction.js]: I didn't realize I could do that or else I would have.

File: test/snapp_auction.js (hunk 2 above)
Comment (bh2smith) [reply test/snapp_auction.js]: Apparently not.

File: test/snapp_auction.js (hunk 3 above)
Comment (bh2smith) [reply test/snapp_auction.js]: Weird. I thought this was a draft.

File: test/utilities.js (hunk 4 above)
Comment (bh2smith) [reply test/utilities.js]: Actually, I would prefer to call this `partitionArray`. 

Does your one liner actually do the same thing? I wouldn't mind using that (but only if I can understand it).

File: test/snapp_auction.js (hunk 2 above)
Comment (fleupold) [reply test/snapp_auction.js]: Can you explain what we are awaiting on in the current version? As far as I can tell forEach returns `void` and await only does something when it is called on a Promise.

I think we have to await on `instance.placeSellOrders(concatenated_orders, { from: user_1 })` inside the forEach function (and declare that one as async).

You can if your current version is working by simulating a long transaction time 
```js
const waitFor = (ms) => new Promise(r => setTimeout(r, ms));
await partitionedOrders.forEach(async function(partition) {  // should be about 10
        concatenated_orders = Buffer.concat(partition)
        await waitFor(100)
        instance.placeSellOrders(concatenated_orders, { from: user_1 })
}
```

and check if your test still passes.

File: test/snapp_auction.js (hunk 1 above)
Comment (fleupold) [reply test/snapp_auction.js]:
```js
await partitionedOrders.forEach(async function(partition) {  // should be about 10
        const concatenated_orders = Buffer.concat(partition)
        await instance.placeSellOrders(concatenated_orders, { from: user_1 })
}
```

File: test/utilities.js (hunk 4 above)
Comment (fleupold) [reply test/utilities.js]: partitionArray sounds good to me. And screw the one liner, i usually prefer easy to read code as well.

**Review by josojo — APPROVED** [review]
Now, it is sparkling clean.
Great

**Pull request merged**